File: content/tracks/dsa/problems/lc-0875-koko-eating-bananas/solution.go
package main

func minEatingSpeed(piles []int, h int) int {
	// The answer lies in [1, max(piles)]: eating faster than the largest pile saves nothing.
	lo, hi := 1, 1
	for _, pile := range piles {
		hi = max(hi, pile)
	}
	for lo < hi {
		mid := lo + (hi-lo)/2
		if hoursAt(piles, mid) <= h {
			hi = mid // mid is fast enough; a slower speed might be too
		} else {
			lo = mid + 1
		}
	}
	return lo
}

func hoursAt(piles []int, speed int) int {
	// Each pile takes ceil(pile / speed) hours.
	hours := 0
	for _, pile := range piles {
		hours += (pile + speed - 1) / speed
	}
	return hours
}

File: content/tracks/dsa/problems/lc-0875-koko-eating-bananas/solution.py
from typing import List


class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        def hours_at(speed: int) -> int:
            # Each pile takes ceil(pile / speed) hours.
            return sum((pile + speed - 1) // speed for pile in piles)

        # The answer lies in [1, max(piles)]: eating faster than the largest pile saves nothing.
        lo, hi = 1, max(piles)
        while lo < hi:
            mid = lo + (hi - lo) // 2
            if hours_at(mid) <= h:
                hi = mid  # mid is fast enough; a slower speed might be too
            else:
                lo = mid + 1
        return lo
